src/Driver.cpp

```diff
@@ -184,3 +165,228 @@ void Driver::startAcquisition()
+{
+    sendStandardCommand(std::string("WP"), pingsPerEnsemble, 5);
+}
+
+void Driver::setDepthCellSize(int depthCellSize)
+{
+    sendStandardCommand(std::string("WS"), depthCellSize, 4);
+}
+
+void Driver::applyConfig(const dvl_teledyne::Config& conf)
+{
+    setBottomTrackPingsPerEnsemble(conf.bottomTrackPingsPerEnsemble);
+    setMaximumTrackingDepth(conf.maximumTrackingDepth);
+    setFlowControlSettings(conf.automaticEnsembleCycling,
+                                conf.automaticPingCycling,
+                                conf.binaryDataOutput,
+                                conf.enableSerialOutput,
+                                conf.enableDataRecording);
+    setHeadingAlignment(conf.headingAlignment);
+    setSalinity(conf.salinity);
+    set_e_headingBias(conf.e_headingBias);
+
+    OutputConfiguration outputConf;
+    outputConf.coordinate_system = conf.transformation;
+    outputConf.use_attitude = conf.useTiltsInTransformation;
+    outputConf.use_3beam_solution = conf.allow3BeamSolutions;
+    outputConf.use_bin_mapping = conf.allowBinMapping;
+    setOutputConfiguration(outputConf);
+
+    setSensorSourceSettings(conf.speedOfSoundSource,
+                                    conf.depthSource,
+                                    conf.headingSource,
+                                    conf.pitchAndRollSource,
+                                    conf.salinitySource,
+                                    conf.temperatureSource);
+    setTimePerEnsemble(conf.timePerEnsemble);
+    setTimeBetweenPings(conf.timeBetweenPings);
+    setNumberOfDepthCells(conf.numberOfDepthCells);
+    setPingsPerEnsemble(conf.pingsPerEnsemble);
+    setDepthCellSize(conf.depthCellSize);
+
+    // Save in case communication fails after altering serial port control settings.
+    writePacket(reinterpret_cast<uint8_t const*>("CK\n"), 3, 100);
+
+    setSerialPortControlSettings(conf.baudrate, conf.parity, conf.stopBits);
+    // Save again to save serialPortControlSettings.
+    writePacket(reinterpret_cast<uint8_t const*>("CK\n"), 3, 100);
+}
+
+void Driver::sendStandardCommand(const std::string& characters, int value, int numDigits, bool sign, bool expert)
+{
+    setConfigurationMode();
+    std::string command = characters;
+
+    if (sign && value < 0)
+        command += "-";
+    else if (sign)
+        command += "+";
+
+    for (int n = getNumberOfDigits(value); n < numDigits; ++n)
+    {
+        command += "0";
+    }
+
+    command += std::to_string(value);
+
+    if (sign)
+        ++numDigits;
+    if (expert)
+        ++numDigits;
+
+    writePacket(reinterpret_cast<uint8_t const*>((command + "\n").c_str()), numDigits + 3);
+    readConfigurationAck();
+}
+
+int Driver::getNumberOfDigits(int value) const
+{
+    int numDigits = 1;
+
+    while (value != 0)
+    {
+        value /= 10;
+        ++numDigits;
+    }
+
+    return numDigits;
+}
+
+std::vector<int> Driver::splitTime(const base::Time& microseconds) const
```

Comment:
You could use time.toString() and then extract the values from there.

Reply:
:-1: there's no guarantee on the format of toString. What he does is the right thing and/or he could have also added the corresponding methods to base::Time itself.

Reply:
> But why is the format not guaranteed?

As in "if someone wants to change the format of toString, there is nothing forbidding him to do so". There is no canonical way to display time as a string. For instance, at some point, somebody added milliseconds (and/or microseconds, as far as I remember).

Reply:
Generally speaking, don't use a conversion to string to convert what's essentially a numerical value to another numerical value. Conversions to string are usually meant for user consumption, not data transfer.

Reply:
You might miss here, that you can use toString (and fromString) with arguments to specify the format you generate/read from?

```
 std::string toString(base::Time::Resolution resolution = Microseconds, const std::string& mainFormat = "%Y%m%d-%H:%M:%S") const;

Reply:
I did miss that, yes. The "format is not consistent" argument is therefore moot. The "it's seriously ugly" argument still stands though.